tizen-api/src/tizen-api-pipeline.c

```diff
@@ -0,0 +1,469 @@
+          }
+        } else {
+          elem->sink = NULL;    /* It is not valid */
+          /** @todo What if it keeps being "NULL"? Exception handling at 2nd frame? */
+        }
+      }
+    }
+  }
+
+  g_assert (gst_buffer_get_size (b) == total_size);
+  if (elem->size > 0)
+    g_assert (gst_buffer_get_size (b) == elem->size);
+
+  /* Iterate e->handles, pass the data to them */
+  for (l = elem->handles; l != NULL; l = l->next) {
+    nns_sink *sink = l->data;
+    nns_sink_cb callback = sink->cb;
+
+    callback (buf, size, &elem->tensorsinfo, sink->pdata);
+
+    /** @todo Measure time. Warn if it takes long. Kill if it takes too long. */
+  }
+
+  g_mutex_unlock (&elem->lock);
+
+  for (i = 0; i < num_mems; i++) {
+    gst_memory_unmap (mem[i], &info[i]);
+  }
+
+  return;
+}
+
+/**
+ * @brief Construct the pipeline (more info in tizen-api.h)
+ */
+int
+nns_pipeline_construct (const char *pipeline_description, nns_pipeline_h * pipe)
+{
+  GError *err = NULL;
+  GstElement *pipeline;
+  GstIterator *it = NULL;
+  int ret = NNS_ERROR_NONE;
+
+  nns_pipeline *pipe_h;
+
+  if (FALSE == gst_init_check (NULL, NULL, &err)) {
+    if (err) {
+      dlog_print (DLOG_ERROR, "nnstreamer-capi-pipeline",
+          "Gstreamer has the following error: %s", err->message);
+      g_error_free (err);
+    } else {
+      dlog_print (DLOG_ERROR, "nnstreamer-capi-pipeline",
+          "Cannot initialize gstreamer. Unknown reason.");
+    }
+    return NNS_ERROR_PIPELINE_FAIL;
+  }
+
+  pipeline = gst_parse_launch (pipeline_description, &err);
+  if (pipeline == NULL) {
+    if (err) {
+      dlog_print (DLOG_ERROR, "nnstreamer-capi-pipeline",
+          "Cannot parse and launch the given pipeline = [%s], %s",
+          pipeline_description, err->message);
+      g_error_free (err);
+    } else {
+      dlog_print (DLOG_ERROR, "nnstreamer-capi-pipeline",
+          "Cannot parse and launch the given pipeline = [%s], unknown reason",
+          pipeline_description);
+    }
+    return NNS_ERROR_PIPELINE_FAIL;
+  }
+
+  pipe_h = g_new0 (nns_pipeline, 1);
+  *pipe = pipe_h;
+  g_assert (GST_IS_PIPELINE (pipeline));
+  pipe_h->element = pipeline;
+  g_mutex_init (&pipe_h->lock);
+  g_mutex_lock (&pipe_h->lock);
+
+  pipe_h->namednodes = g_hash_table_new (g_str_hash, g_str_equal);
+
+  it = gst_bin_iterate_elements (GST_BIN (pipeline));
+  if (it != NULL) {
+    gboolean done = FALSE;
+    GValue item = G_VALUE_INIT;
+    GObject *obj;
+    gchar *name;
+
+    /* Fill in the hashtable, "namednodes" with named Elements */
+    while (!done) {
+      switch (gst_iterator_next (it, &item)) {
+        case GST_ITERATOR_OK:
+          obj = g_value_get_object (&item);
+
+          if (GST_IS_ELEMENT (obj)) {
+            GstElement *elem = GST_ELEMENT (obj);
+            name = gst_element_get_name (elem);
+            if (name != NULL) {
+              element *e = NULL;
+              GstElementFactory *valve = gst_element_factory_find ("valve");
```

Comment:
GstElementFactory * (valve / selector / tensor-sink / app-src)
gst_element_factory_get_element_type()

we can get all the element type out-side the loop.

Reply:
Ah.. great!!